### DISCUSSION TITLE
How to receive a single file as opposed to a form?

### DISCUSSION THREAD

**Discussion by astoff** [fastapi/fastapi#10428] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, UploadFile

app = FastAPI()

@app.post("/upload")
async def receive_file(single_file: UploadFile):
    return {"file_size": len(single_file.file.read())}
```

### Description

In the above example, the endpoint expects an HTML form, so for instance one would have to call

```
curl --form single_file=@picture.jpg http://localhost:8080/receive_file
```

How can it be changed so the endpoint expects the file content directly in the request body?  The matching curl call would be 

```
curl --upload-file picture.jpg http://localhost:8080/receive_file
```

### Operating System

Linux

### FastAPI Version

0.103.1

### Pydantic Version

2.4.2

### Python Version

3.11.5

**Comment by jgould22** [comment, marked as answer]
Something like 
```python
from fastapi import FastAPI, Request

app = FastAPI()
 
@app.put("/upload")
async def receive_file(request: Request):
     data: bytes = await request.body()
     return {"file_size": len(data)}
```

**Reply by astoff** [reply]
Okay, this actually works. Unfortunately it confuses the OpenAPI docs endpoint, which doesn't show the right documentation or offer a file selection button. `Annotated[Request, File()]` doesn't work... is there any way around this?

**Reply by jgould22** [reply]
Yes, when you use the Request directly like that fastapi doesnt know anything about it since it uses types to generate the openapi schema.

You will likely need to define the schema yourself sorta like what is described here https://fastapi.tiangolo.com/advanced/path-operation-advanced-configuration/#custom-openapi-path-operation-schema

**Reply by astoff** [reply]
Thanks for the pointer.
